### PR TITLE
Fix bugs in materialization of the debug info section in MCCAS

### PR DESCRIPTION
**Pull request by rastogishubham**
MCCAS debug info materialization is handled by reading from two buffers, the DistinctDataRef and the CURef. There are a few bugs due to assumptions made when writing this part which are fixed with this patch.

Namely:

1. The assumption that no ULEB or SLEB encoded form will be in the DistinctDataRef when materializing.

2. The assumption that a CU Die will always have a Form in it that doesn't belong in the FormsToPartition.

These assumptions lead to bugs when a situation arises where they do not hold true and have been fixed with this patch.

Another fix is the fact that we do not want to read the bytes form the AbbrevTag but read the ULEB encoded version of the AbbrevTag when writing out the AbbrevTag into the final debug info section CU

There are also some asserts that have been added for safety and sanity checking

### CODE REVIEW

**Comment by rastogishubham** [comment]
@swift-ci please test

**Review by felipepiovezan — APPROVED** [review]
This LGTM!
It bothers me a bit that we don't have a way to unit test this, but I can't come up with a solution for it

**Comment by rastogishubham** [comment]
@felipepiovezan I agree, a unit test isn't feasible because we want to keep the tests small and quick and this stuff is only apparent for large object files, but we have to endure for now, we just have to make sure before we merge a patch, we test clang with clang locally.

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 1]
```diff
@@ -946,8 +946,11 @@ static Error materializeCUDie(DWARFCompileUnit &DCU,
   DWARFDataExtractor DWARFExtractor(CUData, DCU.isLittleEndian(),
                                     DCU.getAddressByteSize());
   uint64_t PrevOffset = CUOffset;
-  uint64_t AbbrevTag = DWARFExtractor.getULEB128(&CUOffset);
-  std::memcpy(&SectionContents[SectionOffset], &AbbrevTag,
+  Error Err = Error::success();
+  uint64_t AbbrevTag = DWARFExtractor.getULEB128(&CUOffset, &Err);
+  if (Err)
+    return Err;
+  std::memcpy(&SectionContents[SectionOffset], &CUData.data()[PrevOffset],
```
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Nit: if `&CUData.data()[PrevOffset]` is equivalent to `CUData.data() + PrevOffset` that might be easier to read.

**Review by adrian-prantl — APPROVED** [review]

**Comment by rastogishubham** [comment]
@swift-ci please test

**Pull request merged**